From 40aca7af22c3406cff0b2ce317e3f9aa7e4939d5
From: Fedor <fedork@ubuntu.(none)>
Date: Tue, 14 Dec 2010 20:31:26 -0500
Subject: [PATCH] [IGMP]: Optimize igmp_rcv

 kernel.org commits:
 cd557bc1c15cbd20fbea47a150e1c7e56834e627 [IGMP]: Optimize kfree_skb in igmp_rcv
 9a57a9d291980302b4a3184fbc47dbddac71903e igmp: avoid two atomic ops in igmp_rcv()

---
 net/ipv4/igmp.c |   11 ++++-------
 1 files changed, 4 insertions(+), 7 deletions(-)

diff --git a/net/ipv4/igmp.c b/net/ipv4/igmp.c
--- a/net/ipv4/igmp.c
+++ b/net/ipv4/igmp.c
@@ -934,17 +934,16 @@ static void igmp_heard_query(struct in_device *in_dev, struct sk_buff *skb,
 	read_unlock(&in_dev->mc_list_lock);
 }
 
+/* called in rcu_read_lock() section */
 int igmp_rcv(struct sk_buff *skb)
 {
 	/* This basically follows the spec line by line -- see RFC1112 */
 	struct igmphdr *ih;
-	struct in_device *in_dev = in_dev_get(skb->dev);
+	struct in_device *in_dev = __in_dev_get_rcu(skb->dev);
 	int len = skb->len;
 
-	if (in_dev==NULL) {
-		kfree_skb(skb);
-		return 0;
-	}
+	if (in_dev == NULL)
+		goto drop;
 
 	if (!pskb_may_pull(skb, sizeof(struct igmphdr)))
 		goto drop;
@@ -977,7 +976,6 @@ int igmp_rcv(struct sk_buff *skb)
 		break;
 	case IGMP_PIM:
 #ifdef CONFIG_IP_PIMSM_V1
-		in_dev_put(in_dev);
 		return pim_rcv_v1(skb);
 #endif
 	case IGMPV3_HOST_MEMBERSHIP_REPORT:
@@ -992,7 +990,6 @@ int igmp_rcv(struct sk_buff *skb)
 	}
 
 drop:
-	in_dev_put(in_dev);
 	kfree_skb(skb);
 	return 0;
 }
-- 
1.6.5.GIT
